src/graph_generator.cc

```diff
@@ -74,25 +115,58 @@ namespace detail {
 			t->transform_task(*tsk, cdag);
 		}
 
+		// Only execution tasks can have data requirements or reductions
+		if(tsk->get_execution_target() != execution_target::NONE) {
 #ifndef NDEBUG
-		// It is currently undefined to split reduction-producer tasks into multiple chunks on the same node:
-		//   - Per-node reduction intermediate results are stored with fixed access to a single SYCL buffer, so multiple chunks on the same node will race on
-		//     this buffer access
-		//   - Inputs to the final reduction command are ordered by origin node ids to guarantee bit-identical results. It is not possible to distinguish
-		//     more than one chunk per node in the serialized commands, so different nodes can produce different final reduction results for non-associative
-		//     or non-commutative operations
-		if(!tsk->get_reductions().empty()) {
-			std::unordered_set<node_id> producer_nids;
-			for(auto& cmd : cdag.task_commands(tid)) {
-				assert(producer_nids.insert(cmd->get_nid()).second);
+			// It is currently undefined to split reduction-producer tasks into multiple chunks on the same node:
+			//   - Per-node reduction intermediate results are stored with fixed access to a single SYCL buffer, so multiple chunks on the same node will race
+			//   on
+			//     this buffer access
+			//   - Inputs to the final reduction command are ordered by origin node ids to guarantee bit-identical results. It is not possible to distinguish
+			//     more than one chunk per node in the serialized commands, so different nodes can produce different final reduction results for non-associative
+			//     or non-commutative operations
+			if(!tsk->get_reductions().empty()) {
+				std::unordered_set<node_id> producer_nids;
+				for(auto& cmd : cdag.task_commands(tid)) {
+					assert(producer_nids.insert(cmd->get_nid()).second);
+				}
 			}
-		}
 #endif
 
-		// TODO: At some point we might want to do this also before calling transformers
-		// --> So that more advanced transformations can also take data transfers into account
-		process_task_data_requirements(tid);
-		process_task_side_effect_requirements(tid);
+			// TODO: At some point we might want to do this also before calling transformers
+			// --> So that more advanced transformations can also take data transfers into account
+			process_task_data_requirements(tid);
+			process_task_side_effect_requirements(tid);
+		}
+
```

Comment:
This particular function getting even larger is not great. I didn't look at it in detail but I think the following `for` and `if` blocks don't have a lot of local dependencies, so maybe those could stand to be a function each?

Reply:
That was in part deliberate: Horizon commands were generated with a function call followed by an early return before, but since we want to be able to attach buffer dependencies on more than just execution commands in the future (#94), all tasks should follow the same overall control flow here.

I agree that this might benefit from subdivision into smaller member functions though.

Reply:
I've now split this into 6 functions.